components/server/src/workspace/gitpod-server-impl.ts

```diff
@@ -1713,6 +1702,35 @@ export class GitpodServerImpl implements GitpodServerWithTracing, Disposable {
         }
     }
 
+    public async setProjectConfiguration(ctx: TraceContext, projectId: string, configString: string): Promise<void> {
+        traceAPIParams(ctx, { projectId }); // filter configString because of size
+
+        const user = this.checkAndBlockUser("setProjectConfiguration");
+        await this.guardProjectOperation(user, projectId, "update");
+        const parseResult = this.gitpodParser.parse(configString);
+        if (parseResult.validationErrors) {
+            throw new Error(`This configuration could not be parsed: ${parseResult.validationErrors.join(', ')}`);
+        }
+        await this.projectsService.updateProjectPartial({
+            id: projectId,
+            config: { '.gitpod.yml': configString },
+        });
+    }
+
+    public async updateProjectPartial(ctx: TraceContext, partialProject: PartialProject): Promise<void> {
+        const user = this.checkUser("updateProjectPartial");
```

Comment:
nit: let's add a `traceAPIParams` like so
```suggestion
        traceAPIParams(ctx, { partialProject: { id: partialProject.id, settings: partialProject.settings });   // censor everything irrelevant

        const user = this.checkUser("updateProjectPartial");
```